conversant/prompt_chatbot.py

```diff
@@ -186,11 +188,103 @@ def _update_max_context_examples(
             "tokens"
         )
 
+    def should_stop(self, response:str)-> bool:
+        """Given a response, decides if it ends with a stop sequence and should 
+        be terminated."""
+        for stop_seq in self.client_config["stop_sequences"]: #make into function
+            if response.endswith(stop_seq):
+                return True
+        return False
+    
+    def get_max_reruns(self):
+        """Returns the maximum number of times partial_reply should be invoked."""
+        return int(self.client_config["max_tokens"]/TOKENS_PER_REQUEST)
+    
+    def partial_reply(self, query: str, is_from_scratch: bool):
```

Comment:
[design commentary] returning the value + a bool here totally works. At the same time, I wonder if it may be cleaner to use a generator here by having the `partial_reply()` use `yield` to get a part of the utterance each time it is called? 

The very rough idea would be that within the `partial_reply()` function, you could have a loop that calls `co.generate()` up to the max amount of reruns. Each time you `yield` the snippet of the utterance generated. If you hit the stop sequence at any point, or the loop runs out of retries, you then can store all the necessary data from the completed `response` in the bot's various history objects.

I _think_ that with this type of design, you could remove the need for the extra boolean that gets returned and the is_from_scratch arg.

When you call `partial_reply()` in the streamlit (or anywhere else external to the chatbot) if using this method, then I think the interface then becomes something like:

```python
for chunk in bot.partial_reply():
  # update state of rendered reply here.
  # this loop should auto-exhaust when partial_reply() stops yielding
  # due to hitting stop sequence within the function and its inner co.generate() loop.

# do stuff after fully rendered reply here
```

FWIW I haven't spent a huge amount of time thinking about this so please lmk if this design either doesn't work and/or does not make the interface more intuitive compared to the interface that is currently here :)

Reply:
Since we are using experimental_rerun, which executes the whole code from scratch, I do not think `yield` will work here. Also, I tried rendering the reply in a for loop and st_chat library we are using doesn't allow updating the component after it is drawn with the same key. I think we can solve that issue, but it will probably be more complicated in the end. 🧐 I wonder what @yichern thinks about this since he is more accustomed to the components 👀

Reply:
I think the yield approach is more intuitive for anyone using this library, and in general I'd be hesitant to optimize specifically for the quirks of Streamlit and our message rendering component. More comments on overall design outside this thread.

Reply:
Thank you for the comments, they are great! Do you think that we should keep working with Streamlit if we want to get a more responsive UI? I might be wrong because it is my first time working with it, but due to the way it is structured it seems like it will result in slower response times, which does not work well with token streaming. BTW, I would be happy to schedule a call (fr this time haha) to chat about your thoughts on how we should tackle this in general or with streamlit! I do not have much experience regarding frontend, so I would really appreciate it! 

Note: Because we are giving the whole chat history as a prompt every time and it picks up where it was left off, I am not sure if the responses would be that different with smaller token sizes (for most of the cases at least) but I might be mistaken here 🤔